src/main/java/org/disq_bio/disq/HtsjdkReadsRddStorage.java

```diff
@@ -0,0 +1,280 @@
+import org.disq_bio.disq.impl.formats.sam.AnySamSinkMultiple;
+import org.disq_bio.disq.impl.formats.sam.SamFormat;
+import org.disq_bio.disq.impl.formats.sam.SamSink;
+import org.disq_bio.disq.impl.formats.sam.SamSource;
+
+/** The entry point for reading or writing a {@link HtsjdkReadsRdd}. */
+public class HtsjdkReadsRddStorage {
+
+  /** An option for configuring how to write a {@link HtsjdkReadsRdd}. */
+  public interface WriteOption {}
+
+  /** An option for configuring which format to write a {@link HtsjdkReadsRdd} as. */
+  public enum FormatWriteOption implements WriteOption {
+    /** BAM format */
+    BAM,
+    /** CRAM format */
+    CRAM,
+    /** SAM format */
+    SAM;
+  }
+
+  /** An option for configuring the number of files to write a {@link HtsjdkReadsRdd} as. */
+  public enum FileCardinalityWriteOption implements WriteOption {
+    /** Write a single file specified by the path. */
+    SINGLE,
+    /** Write multiple files in a directory specified by the path. */
+    MULTIPLE
+  }
+
+  /**
+   * An option for controlling which directory to write temporary part files to when writing a
+   * {@link HtsjdkReadsRdd} as a single file.
+   */
+  public static class TempPartsDirectoryWriteOption implements WriteOption {
+    private String tempPartsDirectory;
+
+    public TempPartsDirectoryWriteOption(String tempPartsDirectory) {
+      this.tempPartsDirectory = tempPartsDirectory;
+    }
+
+    String getTempPartsDirectory() {
+      return tempPartsDirectory;
+    }
+  }
+
+  private JavaSparkContext sparkContext;
+  private int splitSize;
+  private ValidationStringency validationStringency = ValidationStringency.DEFAULT_STRINGENCY;
+  private boolean useNio;
+  private String referenceSourcePath;
+
+  /**
+   * Create a {@link HtsjdkReadsRddStorage} from a Spark context object.
+   *
+   * @param sparkContext the Spark context to use
+   * @return a {@link HtsjdkReadsRddStorage}
+   */
+  public static HtsjdkReadsRddStorage makeDefault(JavaSparkContext sparkContext) {
+    return new HtsjdkReadsRddStorage(sparkContext);
+  }
+
+  private HtsjdkReadsRddStorage(JavaSparkContext sparkContext) {
+    this.sparkContext = sparkContext;
+  }
+
+  /**
+   * @param splitSize the requested size of file splits when reading
+   * @return the current {@link HtsjdkReadsRddStorage}
+   */
+  public HtsjdkReadsRddStorage splitSize(int splitSize) {
+    this.splitSize = splitSize;
+    return this;
+  }
+
+  /**
+   * @param validationStringency the validation stringency for reading
+   * @return the current {@link HtsjdkReadsRddStorage}
+   */
+  public HtsjdkReadsRddStorage validationStringency(ValidationStringency validationStringency) {
+    this.validationStringency = validationStringency;
+    return this;
+  }
+
+  /**
+   * @param useNio whether to use NIO or the Hadoop filesystem (default) for file operations
+   * @return the current {@link HtsjdkReadsRddStorage}
+   */
+  public HtsjdkReadsRddStorage useNio(boolean useNio) {
```

Comment:
Better than this, to allow possible extensibility, I would suggest to convert this method to `useFileSystem(FileSystemWrapper fileSystem)`  and in line 146 use directly that instance.

Reply:
`FileSystemWrapper` is an internal implementation that I'd rather not make a part of the public interface unless there's a good reason to do so. However, internally it's OK to pass around a `FileSystemWrapper` rather than a `useNio` flag, so've made that change.

Reply:
The main reason I suggested that is to use custom file-systems, that are not NIO or Hadoop based. For example, for variants it could be a GenomicsDB. What I'm not sure about is if this is better implemented as a writing-option or the file-system wrapper.